internal/oci/container.go

```diff
@@ -347,18 +390,108 @@ func (c *Container) exitFilePath() string {
 // IsRunning is a function that checks if a specific pid is running
 // it is used to check a container state when we don't want (or don't trust) a `$runtime state` call
 func (c *Container) IsRunning() bool {
-	process, err := os.FindProcess(c.state.Pid)
+	_, err := c.pid()
 	if err != nil {
-		logrus.Infof("container %s not found: %v", c.id, err)
+		logrus.Errorf("checking if pid of %s is running failed: %v", c.id, err)
 		return false
 	}
 
-	if err := process.Signal(syscall.Signal(0)); err != nil {
-		logrus.Errorf("container %s not running %v", c.id, err)
-		return false
+	return true
+}
+
+// Pid is the external, locking function that returns the container's
+// pid, if it's running, and actually the container process
+func (c *Container) Pid() (int, error) {
+	c.opLock.Lock()
+	defer c.opLock.Unlock()
+	return c.pid()
+}
+
+// pid is the internal, non-locking version of pid. It checks
+// that we have a Pid defined in the state, that pid can be found
+// and it is the same process that was originally started by the runtime
+func (c *Container) pid() (int, error) {
+	if c.state == nil {
+		return -1, errors.New("state not initialized")
+	}
+	pid := c.state.InitialPid
+	if pid <= 0 {
+		return -1, errors.New("pid not initialized")
 	}
 
-	return true
+	// container has stopped (as pid is initialized but the runc state has overwritten it)
+	if c.state.Pid == 0 {
+		return 0, findprocess.ErrNotFound
+	}
+
+	if _, err := findprocess.FindProcess(pid); err != nil {
+		return -1, err
+	}
+
+	if err := c.verifyPid(pid); err != nil {
+		return -1, err
+	}
+	return pid, nil
+}
+
+// findAndReleasePid attempts to find the container's pid
+// and if it's found, release it. In between these steps,
+// it also verified the pid running is the same as was originally
+// started by the runtime
+func (c *Container) findAndReleasePid() (bool, error) {
+	pid := c.state.InitialPid
+	found := false
```

Comment:
IMHO using this variable in the code below decreases the readability, as you can use `return true` (or `false`) directly.